apps/volontulo/models.py

```diff
@@ -38,13 +111,175 @@ class Offer(models.Model):
     votes = models.BooleanField(default=0)
+    recruitment_start_date = models.DateTimeField(blank=True, null=True)
+    recruitment_end_date = models.DateTimeField(blank=True, null=True)
+    reserve_recruitment = models.BooleanField(blank=True, default=True)
+    reserve_recruitment_start_date = models.DateTimeField(
+        blank=True,
+        null=True
+    )
+    reserve_recruitment_end_date = models.DateTimeField(
+        blank=True,
+        null=True
+    )
+    action_ongoing = models.BooleanField(default=False, blank=True)
+    constant_coop = models.BooleanField(default=False, blank=True)
+    action_start_date = models.DateTimeField(blank=True, null=True)
+    action_end_date = models.DateTimeField(blank=True, null=True)
+    volunteers_limit = models.IntegerField(default=0, null=True, blank=True)
 
     def __str__(self):
         return self.title
 
+    def is_visible(self):
+        u"""Determine offer visibility for specified user type."""
+        statuses_keys = (
+            self.offer_status,
+            self.recruitment_status,
+            self.action_status
+        )
+        return True if self.VISIBILITY_MATRIX[statuses_keys] else False
+
+    def set_main_image(self, form):
+        u"""Set main image flag unsetting other offers images.
+
+        :param offer: Offer model instance
+        """
+        if form.cleaned_data["is_main"]:
+            OfferImage.objects.filter(offer=self).update(is_main=False)
+            return True
+        return False
+
+    def save_offer_image(self, form, userprofile):
+        u"""Handle image upload for user profile page.
+
+        :param offer: Offer model instance
+        """
+        if form.is_valid():
+            gallery = form.save(commit=False)
+            gallery.offer = self
+            gallery.userprofile = userprofile
+            gallery.is_main = self.set_main_image(form)
+            gallery.save()
+            return True
+        else:
+            return form.errors
+
+    def create_new(self):
+        u"""Set status while creating new offer."""
+        self.offer_status = 'unpublished'
+        self.recruitment_status = 'open'
+        self.action_status = self.determine_action_status()
+
+    def determine_action_status(self):
+        u"""Determine action status by offer dates."""
+        if (
+                (
+                    self.finished_at and
+                    self.started_at < timezone.now() < self.finished_at
+                ) or
+                (
+                    self.started_at < timezone.now() and
+                    not self.finished_at
+                )
+        ):
+            return 'ongoing'
+        elif self.started_at > timezone.now():
+            return 'future'
+        else:
+            return 'finished'
+
+    def change_status(self, request):
+        u"""Change offer status.
+
+        :param offer: Offer model instance
+        """
+        if request.POST.get('status') in {'published', 'rejected'}:
```

Comment:
It's not dictionary just a set and I think it's faster in `x in {d}` expression.